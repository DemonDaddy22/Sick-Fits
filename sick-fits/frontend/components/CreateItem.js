import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import Error from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

export const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ) {
        createItem(
            data: {
                title: $title
                description: $description
                price: $price
                image: $image
                largeImage: $largeImage
            }
        ) {
            id
        }
    }
`;

export default class CreateItem extends React.Component {

    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0
    }

    handleChange = e => {
        const { name, type, value } = e.target;
        const val = type === 'number' && val ? parseFloat(value) : value;
        this.setState({ [name]: val });
    }

    handleFileUpload = async e => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'sickfits');
        const res = await fetch('https://api.cloudinary.com/v1_1/yelp-camp/image/upload', {
            method: 'POST',
            body: data
        });
        const file = await res.json();
        this.setState({ image: file.secure_url, largeImage: file.eager[0].secure_url });
    }

    handleSubmit = async (e, createItem) => {
        e.preventDefault();
        const res = await createItem();
        Router.push({
            pathname: '/item',
            query: { id: res.data.createItem.id }
        })
    }

    render = () => {
        const { title, description, price, image } = this.state;

        return <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
            {(createItem, { loading, error }) => <Form onSubmit={e => this.handleSubmit(e, createItem)}>
                <Error error={error} />
                <fieldset disabled={loading} aria-busy={loading}>
                    {/* TODO - make input fields bigger and adjust margins */}
                    <label htmlFor='image'>Image</label>
                    <input accept='image/*' type='file' id='image' name='image' placeholder='Upload an Image'
                        onChange={this.handleFileUpload} required />
                    {image && <img width={200} src={image} alt='Upload Preview' />}
                    <label htmlFor='title'>Title</label>
                    <input type='text' id='title' name='title' placeholder='Enter a Title'
                        value={title} onChange={this.handleChange} required />
                    <label htmlFor='price'>Price</label>
                    <input type='number' id='price' name='price' placeholder='Enter a Price'
                        value={price} onChange={this.handleChange} required />
                    <label htmlFor='description'>Description</label>
                    <textarea id='description' name='description' placeholder='Enter a Description'
                        value={description} onChange={this.handleChange} required />
                    <button type='submit'>Submit</button>
                </fieldset>
            </Form>}
        </Mutation>;
    }
}