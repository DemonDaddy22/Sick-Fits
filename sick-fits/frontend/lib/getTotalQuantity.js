const getTotalQuantity = cart => !cart || !Array.isArray(cart) || (Array.isArray(cart) && cart.length === 0)
    ? 0
    : cart.reduce((acc, curr) => acc + curr.quantity, 0);

export default getTotalQuantity;