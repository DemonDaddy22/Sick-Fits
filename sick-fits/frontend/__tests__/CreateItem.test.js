import { mount } from 'enzyme';
import Router from 'next/router';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeItem } from '../lib/testUtils';

const fooImage = 'https://foo.com/bar.jpg';

global.fetch = jest.fn().mockResolvedValue({
    json: () => ({
        secure_url: fooImage,
        eager: [{ secure_url: fooImage }]
    })
});

describe('<CreateItem />', () => {
    it('renders and matches snapshot', () => {
        const wrapper = mount(<MockedProvider>
            <CreateItem />
        </MockedProvider>);

        const form = wrapper.find('[data-test="createItem"]');
        expect(toJSON(form)).toMatchSnapshot();
    });

    it('uploads a file when changed', async () => {
        const wrapper = mount(<MockedProvider>
            <CreateItem />
        </MockedProvider>);

        const input = wrapper.find('input[type="file"]');
        input.simulate('change', { target: { files: ['foobar.jpg'] } });
        await wait();
        const component = wrapper.find('CreateItem').instance();
        // console.log(component);
        expect(component.state.image).toEqual(fooImage);
        expect(component.state.largeImage).toEqual(fooImage);
        expect(global.fetch).toHaveBeenCalled();
        global.fetch.mockReset();
    });

    it('handles state updating', async () => {
        const wrapper = mount(<MockedProvider>
            <CreateItem />
        </MockedProvider>);

        wrapper.find('#title').simulate('change', { target: { value: 'Testing', name: 'title' } });
        wrapper.find('#price').simulate('change', { target: { value: 50000, name: 'price', type: 'number' } });
        wrapper.find('#description').simulate('change', { target: { value: 'Testing item creation', name: 'description' } });

        expect(wrapper.find('CreateItem').instance().state).toMatchObject({
            title: 'Testing',
            price: 50000,
            description: 'Testing item creation'
        });
    });

    it('creates an item when the form is submitted', async () => {
        const item = fakeItem();
        const mocks = [{
            request: {
                query: CREATE_ITEM_MUTATION,
                variables: {
                    title: item.title,
                    description: item.description,
                    image: '',
                    largeImage: '',
                    price: item.price
                }
            },
            result: {
                data: {
                    createItem: { ...item, id: 'foobar123', __typename: 'Item' }
                }
            }
        }];

        const wrapper = mount(<MockedProvider mocks={mocks}>
            <CreateItem />
        </MockedProvider>);

        wrapper.find('#title').simulate('change', { target: { value: item.title, name: 'title' } });
        wrapper.find('#price').simulate('change', { target: { value: item.price, name: 'price', type: 'number' } });
        wrapper.find('#description').simulate('change', { target: { value: item.description, name: 'description' } });

        Router.router = { push: jest.fn() };
        wrapper.find('form').simulate('submit');

        await wait(50);
        wrapper.update();

        expect(Router.router.push).toHaveBeenCalled();
        expect(Router.router.push).toHaveBeenCalledWith({
            pathname: '/item',
            query: { id: 'foobar123' }
        });
    });
});