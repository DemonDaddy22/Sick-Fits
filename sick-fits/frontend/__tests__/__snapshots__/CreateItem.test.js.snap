// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders and matches snapshot 1`] = `
Array [
  <Form
    data-test="createItem"
    onSubmit={[Function]}
  >
    <form
      className="Form-sc-1xszr8q-0 dUkfCz"
      data-test="createItem"
      onSubmit={[Function]}
    >
      <DisplayError
        error={Object {}}
      />
      <fieldset
        aria-busy={false}
        disabled={false}
      >
        <label
          htmlFor="image"
        >
          Image
        </label>
        <input
          accept="image/*"
          id="image"
          name="image"
          onChange={[Function]}
          placeholder="Upload an Image"
          required={true}
          type="file"
        />
        <label
          htmlFor="title"
        >
          Title
        </label>
        <input
          id="title"
          name="title"
          onChange={[Function]}
          placeholder="Enter a Title"
          required={true}
          type="text"
          value=""
        />
        <label
          htmlFor="price"
        >
          Price
        </label>
        <input
          id="price"
          name="price"
          onChange={[Function]}
          placeholder="Enter a Price"
          required={true}
          type="number"
          value={0}
        />
        <label
          htmlFor="description"
        >
          Description
        </label>
        <textarea
          id="description"
          name="description"
          onChange={[Function]}
          placeholder="Enter a Description"
          required={true}
          value=""
        />
        <button
          type="submit"
        >
          Submit
        </button>
      </fieldset>
    </form>
  </Form>,
  <form
    className="Form-sc-1xszr8q-0 dUkfCz"
    data-test="createItem"
    onSubmit={[Function]}
  >
    <DisplayError
      error={Object {}}
    />
    <fieldset
      aria-busy={false}
      disabled={false}
    >
      <label
        htmlFor="image"
      >
        Image
      </label>
      <input
        accept="image/*"
        id="image"
        name="image"
        onChange={[Function]}
        placeholder="Upload an Image"
        required={true}
        type="file"
      />
      <label
        htmlFor="title"
      >
        Title
      </label>
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Enter a Title"
        required={true}
        type="text"
        value=""
      />
      <label
        htmlFor="price"
      >
        Price
      </label>
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Enter a Price"
        required={true}
        type="number"
        value={0}
      />
      <label
        htmlFor="description"
      >
        Description
      </label>
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a Description"
        required={true}
        value=""
      />
      <button
        type="submit"
      >
        Submit
      </button>
    </fieldset>
  </form>,
]
`;
